core/state.py

```diff
@@ -231,6 +232,31 @@ def __exit__(self, type, value, traceback):
         self.mem.running_err_trap = False
 
 
+class ctx_HideErrTrap(object):
+    """For trap ERR."""
+
+    def __init__(self, trap_state, errtrace):
+        # type: (trap_osh.TrapState, bool) -> None
+        self.errtrace = errtrace
+        if self.errtrace: return
+        # errtrace is disabled, hide err trap
+        self.trap_state = trap_state
+        self.err_handler = trap_state.GetHook('ERR')
+        trap_state.RemoveUserHook('ERR')
+
+    def __enter__(self):
+        # type: () -> None
+        pass
+
+    def __exit__(self, type, value, traceback):
+        # type: (Any, Any, Any) -> None
+        if self.errtrace: return
+        # errtrace is disabled, restore err trap (that was removed earlier)
+        # unless it has been set meanwhile
+        if self.err_handler and not self.trap_state.GetHook('ERR'):
+            self.trap_state.AddUserHook('ERR', self.err_handler)
```

Comment:
There are some boolean checks now in `self._MaybeRunErrTrap()` --- can we use that style instead of a context manager?

Reply:
Not sure, as I found that err trap defined in {} is not restored on }, which make sense. This is why I dedicated a context just for that because of the "unless it has been set meanwhile"... maybe should I said replaced or overwritten?

Reply:
Hm not sure what you mean, though as mentioned I suspect this isn't working right because test case 12 in `builtin-trap-err` still fails